### PR TITLE
ADR - Controller structure

### PR DESCRIPTION
**Pull request by darokel**
### Context

- https://trello.com/c/KSCQ9WJz/3812-adr-for-controller-structure-with-nested-resources

### Changes proposed in this pull request

- Raise ADR to document our approach to structuring controllers for nested resources

### Guidance to review

- Read ADR

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by darokel** [comment]
[ADR for controller structure with nested resources](https://trello.com/c/KSCQ9WJz/3812-adr-for-controller-structure-with-nested-resources)

**Review by asmega — APPROVED** [review]

File: docs/adr/0005-controller-structure.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
+# 5. Controller Structure
```
Comment (misaka) [inline docs/adr/0005-controller-structure.md:1]: Sorry but you'll have to renumber this to `6`, this PR reminded me that I need to finish mine, but it's numbered `5`. ;)

File: docs/adr/0005-controller-structure.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+We cannot design endpoints in a flat structure due to existing requirements around the Find service. By flat structure, we're referring to just having endpoints like:
+
+- `/providers`
+- `/courses`
+
+Instead, we have to design endpoints in a nested structure scoped to their parent resources, ie:
+
+- `/recruitment_cycles/:recruitment_cycle_id/providers/:provider_code`
+- `/recruitment_cycles/:recruitment_cycle_id/providers/:provider_code/courses/:course_code`.
```
Comment (misaka) [inline docs/adr/0005-controller-structure.md:19]:
```suggestion
- `/recruitment_cycles/:recruitment_cycle_year/providers/:provider_code`
- `/recruitment_cycles/:recruitment_cycle_year/providers/:provider_code/courses/:course_code`.
```

File: docs/adr/0005-controller-structure.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+As an example, given the following routes:
+
+```ruby
+  namespace :public do
+    namespace :v1 do
+      resources :recruitment_cycles, param: :year, only: [:show] do
+        resources :courses, only: %i[index]
+        resources :providers, only: %i[index show], param: :code do
+          scope module: :providers do
```
Comment (misaka) [inline docs/adr/0005-controller-structure.md:78]: I had to some work to understand the (small) difference between these routes and the one above. Would it be worth adding a comment above line 78 here to explain that `scope` is necessary to get the directory structure right as below?

Comment (darokel) [reply docs/adr/0005-controller-structure.md:80]: done https://github.com/DFE-Digital/teacher-training-api/pull/1535/commits/930f98ced2c1ede0c1365b3133974c859b39bdfe

File: docs/adr/0005-controller-structure.md (hunk 1 above)
Comment (darokel) [reply docs/adr/0005-controller-structure.md]: Done https://github.com/DFE-Digital/teacher-training-api/pull/1535/commits/2470361600d37f9a5c18de52a10bdb0838be2b88

File: docs/adr/0005-controller-structure.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+## Decision
+
+We have decided to go with option 2 based on the number of benefits we will gain given our current requirements.
+
+## Consequences
+
+- Improved maintainability of the project's source code over a long term period
+- Potentially helps with keeping controller actions RESTful by leveraging modules to communicate behaviour
+
+Given the number of options Rails provides for structuring routes, the use of multiple controlelrs for a given resource is not an uncommon pattern in the wild. See the links below for further reading and examples.
```
Comment (misaka) [inline docs/adr/0005-controller-structure.md:121]:
```suggestion
Given the number of options Rails provides for structuring routes, the use of multiple controllers for a given resource is not an uncommon pattern in the wild. See the links below for further reading and examples.
```

**Review by misaka — APPROVED** [review]
Excellent, nice one @darokel. 💯 

Minor typo, but good to go.

**Pull request merged**